[OPIK-2469] [BE] Fix duplicate experiment items in dataset comparison

Nimrod007 opened the pull request:
## Details

Fixed a bug where duplicate experiment items were appearing in dataset comparison results when traces existed across multiple projects.

### Root Cause: Cross-Project Traces

When the same trace exists in multiple projects (cross-project traces), the SQL query was grouping span aggregations by `project_id`, causing multiple rows to be returned for a single trace. This led to duplicate experiment items in the final result.

**The Problem:**
```sql
-- BUGGY: Grouping by project_id causes duplicates for cross-project traces
LEFT JOIN (
    SELECT 
        trace_id,
        SUM(total_estimated_cost) AS total_estimated_cost,
        sumMap(usage) AS usage
    FROM spans final
    WHERE workspace_id = :workspace_id
    AND trace_id IN (SELECT trace_id FROM experiment_items_scope)
    GROUP BY workspace_id, project_id, trace_id  -- ❌ Returns 2 rows for cross-project traces
) s ON t.id = s.trace_id
```

When a trace has spans in Project A and Project B, the query returns:
- Row 1: `trace_id` with `project_id = A`
- Row 2: `trace_id` with `project_id = B`

This causes the JOIN to duplicate experiment items in the `groupArray()` aggregation.

### Solution: Aggregate Across All Projects

Removed `project_id` from the `GROUP BY` clause in the spans aggregation subquery. This ensures each trace returns exactly one row with aggregated usage metrics across all projects.

```sql
-- FIXED: Aggregate across all projects for each trace
LEFT JOIN (
    SELECT 
        trace_id,
        SUM(total_estimated_cost) AS total_estimated_cost,
        sumMap(usage) AS usage
    FROM spans final
    WHERE workspace_id = :workspace_id
    AND trace_id IN (SELECT trace_id FROM experiment_items_scope)
    GROUP BY workspace_id, trace_id  -- ✅ Returns 1 row per trace
) s ON t.id = s.trace_id
```

**Benefits:**
- Eliminates duplicates for cross-project traces
- Correctly aggregates usage metrics across all projects
- More architecturally sound solution

### Production Verification

Verified the fix using SQL queries on production database:

**Without Fix (Buggy):**
- 7 out of 10 traces returned 2 rows each (cross-project traces)
- API returned 17 experiment items (10 unique + 7 duplicates)

**With Fix:**
- All 10 traces return exactly 1 row
- API returns 10 unique experiment items (41% reduction)

**SQL Verification Query:**
```sql
WITH 
experiment_items_scope AS (
    SELECT * FROM experiment_items final
    WHERE workspace_id = :workspace_id AND experiment_id = :experiment_id
),
buggy_version AS (
    SELECT trace_id, COUNT(*) as row_count
    FROM (
        SELECT trace_id FROM spans final
        WHERE workspace_id = :workspace_id
        AND trace_id IN (SELECT DISTINCT trace_id FROM experiment_items_scope)
        GROUP BY workspace_id, project_id, trace_id
    )
    GROUP BY trace_id
)
SELECT trace_id, row_count FROM buggy_version WHERE row_count > 1;
```

Result: 7 traces with `row_count = 2` confirmed the bug in production.

### Changed Files

- `apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java` (line 692)

## Change checklist

- [x] User facing (bug fix - eliminates duplicate experiment items in UI)
- [ ] Documentation update (internal bug fix, no API changes)

## Issues

- Resolves #
- OPIK-2469

## Testing

### Verification Approach

1. ✅ **Root Cause Analysis**: Identified cross-project traces as the source of duplicates
2. ✅ **SQL Verification**: Ran diagnostic queries on production database confirming 7 traces with duplicates
3. ✅ **Fix Validation**: Confirmed fixed query returns unique rows for all traces
4. ✅ **Customer Data Analysis**: Verified duplicates exist in production environment
5. ✅ **Code Compilation**: All code compiles successfully

### Production Impact

- **Before**: 70% more items returned due to cross-project trace duplicates
- **After**: Only unique experiment items returned
- **Customer Impact**: Duplicate sidebars in dataset comparison UI will be resolved

## Documentation

No documentation updates required - this is an internal bug fix that maintains the existing API contract while fixing the duplicate items issue.

andrescrz commented on apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java:
```diff
@@ -681,14 +681,14 @@ ORDER BY (workspace_id, project_id, id) DESC, last_updated_at DESC
+                FROM spans final
+                WHERE workspace_id = :workspace_id
+                AND trace_id IN (SELECT trace_id FROM experiment_items_scope)
+                GROUP BY workspace_id, trace_id
```
I'm not fully convinced about this summary, specially the 40% gain.

The best way would be by presenting the objective data, in the form of the query plans before and after the change, particularly the queries were the issue was reported.

Anyway, this is more a wish on my end. Not a blocker to continue.

andrescrz reviewed (approved):
Thanks for addressing the comments, no blockers on my end.

Nimrod007 commented:
> Thanks for addressing the comments, no blockers on my end.

@andrescrz 40% is wrong since its tested over a single workspace (which only for that specific case it was 40%) but the general check is correct from my review. that it should be better vs worse
